httpcache.go

```diff
@@ -498,6 +516,41 @@ func headerAllCommaSepValues(headers http.Header, name string) []string {
 	return vals
 }
 
+// cachedReadCloser is a wrapper around ReadCloser R that calls OnEOF
+// handler with a full copy of the content read from that ReadCloser
+// when EOF is read.
+type cachedReadCloser struct {
+	// Underlying ReadCloser.
+	R io.ReadCloser
+	// OnEOF is called with a copy of the content of R when EOF is reached.
+	OnEOF func(io.ReadCloser)
+
+	buf bytes.Buffer // buf stores a copy of the content of r.
+}
+
+// Read reads the next len(p) bytes from r or until r is drained. The
+// return value n is the number of bytes read. If r has no data to
+// return, err is io.EOF (unless len(p) is zero) and onEOF is called
```

Comment:
Also, drop the (unless len(p) is zero) part. It's not helpful.

If len(p) is zero, no data is requested, so there can't be no data to return since we didn't ask for any.

"If r has no data to return" implies len(p) > 0 to me.